src/solvers/refinement/string_refinement.cpp

```diff
@@ -270,12 +270,15 @@ static void make_char_array_pointer_associations(
   }
 }
 
-void replace_symbols_in_equations(
-  const union_find_replacet &symbol_resolve,
-  std::vector<equal_exprt> &equations)
+/// Substitute sub-expressions in equation by representative elements of
+/// `symbol_resolve` whenever possible.
+/// Similar to `symbol_resolve.replace_expr` but doesn't mutate the expression
+/// and return the transformed expression instead.
+static exprt
+transform_expr(const union_find_replacet &symbol_resolve, exprt expr)
```

Comment:
`exprt`-> `equal_exprt` (in the doc you mention `equations`)

Reply:
That's more or less what I had in mind. However `transform_expr` sounds like `do_something_to_expr`. :smile: 
How about `substitute_with_repr`?

Reply:
I would keep `transform` as it is used in the standard library for operations without side-effects https://en.cppreference.com/w/cpp/algorithm/transform

Reply:
Yeah but in `std::transform` you give the operation as argument, so the generic name makes sense there. Not here, because the caller doesn't know what the operation is.

Reply:
Yes you are right, but instead of having a "substitute" and a "replace" function name (difficult to see which one is which from the name), I suggest to name this one `replace_expr_copy` (following the `_copy` pattern of the stl), making it clear that this one creates a new copy of the expression.